engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -255,99 +332,149 @@ private Void handleCreateVolumeFromSnapshotBothOnStorageSystem(SnapshotInfo snap
-        int primaryStorageDownloadWait = NumbersUtil.parseInt(value, Integer.parseInt(Config.PrimaryStorageDownloadWait.getDefaultValue()));
-        CopyCommand copyCommand = new CopyCommand(snapshotInfo.getTO(), volumeInfo.getTO(), primaryStorageDownloadWait, VirtualMachineManager.ExecuteInSequence.value());
+            if (useCloning) {
+                copyCmdAnswer = performResignature(volumeInfo, hostVO);
+            }
+            else {
+                // asking for a XenServer host here so we don't always prefer to use XenServer hosts that support resigning
+                // even when we don't need those hosts to do this kind of copy work
+                hostVO = getXenServerHost(snapshotInfo.getDataCenterId(), false);
 
-        CopyCmdAnswer copyCmdAnswer = null;
+                copyCmdAnswer = performCopyOfVdi(volumeInfo, snapshotInfo, hostVO);
+            }
 
-        try {
-            _volumeService.grantAccess(snapshotInfo, hostVO, snapshotInfo.getDataStore());
-            _volumeService.grantAccess(volumeInfo, hostVO, volumeInfo.getDataStore());
+            if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
+                if (copyCmdAnswer != null && copyCmdAnswer.getDetails() != null && !copyCmdAnswer.getDetails().isEmpty()) {
+                    errMsg = copyCmdAnswer.getDetails();
+                }
+                else {
+                    errMsg = "Unable to perform host-side operation";
+                }
+            }
+        }
+        catch (Exception ex) {
+            errMsg = ex.getMessage() != null ? ex.getMessage() : "Copy operation failed";
+        }
 
-            Map<String, String> srcDetails = getSnapshotDetails(_storagePoolDao.findById(snapshotInfo.getDataStore().getId()), snapshotInfo);
+        CopyCommandResult result = new CopyCommandResult(null, copyCmdAnswer);
 
-            copyCommand.setOptions(srcDetails);
+        result.setResult(errMsg);
 
-            Map<String, String> destDetails = getVolumeDetails(volumeInfo);
+        callback.complete(result);
+    }
 
-            copyCommand.setOptions2(destDetails);
+    // If the underlying storage system is making use of read-only snapshots, this gives the storage system the opportunity to
+    // create a volume from the snapshot so that we can copy the VHD file that should be inside of the snapshot to secondary storage.
+    //
+    // The resultant volume must be writable because we need to resign the SR and the VDI that should be inside of it before we copy
+    // the VHD file to secondary storage.
+    //
+    // If the storage system is using writable snapshots, then nothing need be done by that storage system here because we can just
+    // resign the SR and the VDI that should be inside of the snapshot before copying the VHD file to secondary storage.
+    private void createVolumeFromSnapshot(HostVO hostVO, SnapshotInfo snapshotInfo, boolean keepGrantedAccess) {
```

Comment:
@mike-tutkowski Could you please use Javadoc (`/** <documentation> */`) instead of comments (`// <documentation>`)?
As you already wrote a good documentation, I think it would enrich your work.

Thanks.